code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java

```diff
@@ -69,27 +71,62 @@ public Account getAdminAccount() {
     }
 
     @Override
-    public List<Account> searchAccounts(String name) {
-        List<Account> accounts = new ArrayList<>();
-        accounts.addAll(create()
-                .selectFrom(ACCOUNT)
-                .where(ACCOUNT.STATE.eq(CommonStatesConstants.ACTIVE)
-                        .and(ACCOUNT.NAME.contains(name))
-                        .and(ACCOUNT.KIND.eq(AccountConstants.USER_KIND)
-                                .or(ACCOUNT.KIND.eq(AccountConstants.ADMIN_KIND))))
-                .orderBy(ACCOUNT.ID.asc()).fetch());
-        return accounts;
+    public List<Account> searchUsers(String username) {
+        try {
+            List<Account> accounts = new ArrayList<>();
+            accounts.addAll(create()
+                    .select(ACCOUNT.fields())
+                    .from(ACCOUNT)
+                    .join(CREDENTIAL)
+                    .on(CREDENTIAL.ACCOUNT_ID.eq(ACCOUNT.ID))
+                    .where(ACCOUNT.STATE.eq(CommonStatesConstants.ACTIVE)
+                            .and(CREDENTIAL.STATE.eq(CommonStatesConstants.ACTIVE))
+                            .and(CREDENTIAL.PUBLIC_VALUE.contains(username))
+                            .and(CREDENTIAL.KIND.eq("password")))
+                    .orderBy(ACCOUNT.ID.asc()).fetchInto(AccountRecord.class));
+            return accounts;
+        } catch (InvalidResultException e) {
+            throw new ClientVisibleException(ResponseCodes.CONFLICT, "MultipleSameUsernames");
+        }
     }
 
     @Override
-    public Account getByName(String name) {
-        return create()
-                .selectFrom(ACCOUNT)
-                .where(ACCOUNT.STATE.eq(CommonStatesConstants.ACTIVE)
-                        .and(ACCOUNT.NAME.eq(name))
-                        .and(ACCOUNT.KIND.eq(AccountConstants.USER_KIND)
-                                .or(ACCOUNT.KIND.eq(AccountConstants.ADMIN_KIND))))
-                .orderBy(ACCOUNT.ID.asc()).fetchOneInto(Account.class);
+    public Account getByUsername(String username) {
+        try {
+            return create()
+                    .select(ACCOUNT.fields())
+                    .from(ACCOUNT)
+                    .join(CREDENTIAL)
+                    .on(CREDENTIAL.ACCOUNT_ID.eq(ACCOUNT.ID))
+                    .where(
+                            ACCOUNT.STATE.eq(CommonStatesConstants.ACTIVE)
+                                    .and(CREDENTIAL.STATE.eq(CommonStatesConstants.ACTIVE))
+                                    .and(CREDENTIAL.PUBLIC_VALUE.eq(username)))
+                    .and(CREDENTIAL.KIND.eq("password"))
+                    .fetchOneInto(AccountRecord.class);
+        } catch (InvalidResultException e) {
+            throw new ClientVisibleException(ResponseCodes.CONFLICT, "MultipleOfUsername");
```

Comment:
If it's `inactive` then it's not deleted yet, so that publicValue *should* still be reserved.

@ibuildthecloud I reiterate that it is absurd and dangerous to complicate authentication logic with extra code to handle multiple credentials in multiple states for multiple accounts with the same publicValue.  Here be dragons.  There should be a hard unique on publicValue so there is only one possible entry to authenticate against.

Once a credential deleted/purged that publicValue *could* be eligible for reuse, but that's a feature we don't necessarily need.

 - The simple option is to just not support that, the name is taken forever.
 - If someone *really* wants that name, they can still go manipulate the database.
 - Later we can move the row to a separate non-unique table of deleted credentials for audit-trail/the cloud only scales up.
 - Or you could put in something hacky like on delete update the public value to "*unixTimestamp*oldValue", using a character that we don't allow for API keys or usernames.